File: cursos/pria/src/csharp/7_Bibliotecas.cs
using System;
using System.IO;  // Para trabajar con archivos
using System.Collections.Generic;  // Para trabajar con colecciones de datos
using System.Linq;  // Para realizar operaciones de consulta

class Program
{
    static void Main()
    {
        /////////////////////////////////////////
        // EJEMPLOS DE USAR LIBRERÍAS INTERNAS //
        /////////////////////////////////////////

        // MATEMÁTICAS (FUNCIONES INCORPORADAS)

        float x = 5.0f;
        float y = 2.0f;

        float z = MathF.PI;              // Constante π                              Resultado: 3.141592...
        z = MathF.E;                     // Constante e (Número de Euler)            Resultado: 2.718281...

        z = MathF.Pow(x, y);             // Devuelve x elevado a la potencia de y
        z = MathF.Exp(x);                // Devuelve la función exponencial de un número
        z = MathF.Log(x);                // Devuelve el logaritmo natural de un número

        z = MathF.Abs(x);                // Devuelve el valor absoluto de un número

        z = MathF.Round(x);              // Devuelve el valor de un número redondeado al entero más cercano
        z = MathF.Floor(x);              // Devuelve el entero más grande menor o igual a un número
        z = MathF.Ceiling(x);            // Devuelve el entero más pequeño mayor o igual a un número

        z = MathF.Sin(x);                // Devuelve el seno de un ángulo (ángulo en radianes)
        z = MathF.Cos(x);                // Devuelve el coseno de un ángulo (ángulo en radianes)
        z = MathF.Tan(x);                // Devuelve la tangente de un ángulo (ángulo en radianes)
        z = MathF.Atan(x);               // Devuelve el arcotangente de un número, en radianes
        z = MathF.Atan2(y, x);           // Devuelve el arcotangente del cociente de sus argumentos (y sobre x), en radianes

        z = MathF.Max(x, y);             // Devuelve el mayor de dos números
        z = MathF.Min(x, y);             // Devuelve el menor de dos números

        // ALEATORIEDAD (FUNCIONES INCORPORADAS)

        // Obteniendo un número flotante aleatorio entre 0 y 1
        Random aleatorio = new Random();
        float numeroAleatorio = (float)aleatorio.NextDouble(); // Ejemplo resultado: 0.64816...

        // Obteniendo un número entero aleatorio entre dos valores, inclusivo (ejemplo: entre -3 y 10, inclusivo)
        int min = -3;
        int max = 10;
        int enteroAleatorio = aleatorio.Next(min, max + 1); // El límite superior es exclusivo, por eso agregamos 1

        // COLECCIONES (USANDO System.Collections.Generic)

        // Lista de números de punto flotante
        List<float> listaDeFlotantes = new List<float> { 1.5f, 2.3f, 3.7f, 4.2f, 5.9f };

        // Añadir elementos a la lista
        listaDeFlotantes.Add(6.1f);
        listaDeFlotantes.AddRange(new float[] { 7.4f, 8.6f, 9.0f, 10.3f });

        // Diccionario con claves de tipo string y valores de tipo float
        Dictionary<string, float> diccionarioDePrecios = new Dictionary<string, float>();
        diccionarioDePrecios.Add("Manzana", 0.75f);
        diccionarioDePrecios.Add("Pan", 1.50f);
        diccionarioDePrecios["Leche"] = 2.99f; // Otra forma de añadir elementos al diccionario

        // LINQ (USANDO System.Linq)

        // Filtrar elementos: obtener todos los números mayores que 5
        IEnumerable<float> mayoresQueCinco = listaDeFlotantes.Where(numero => numero > 5f);

        // Ordenar la lista de manera ascendente
        IOrderedEnumerable<float> listaOrdenada = listaDeFlotantes.OrderBy(numero => numero);

        // ESCRIBIR Y LEER ARCHIVOS DE TEXTO (USANDO System.IO)

        string nombreArchivo = "texto.txt";
        string texto = "¡Hola, mundo!" + Environment.NewLine + "¡Esta es otra línea de texto!";

        // Escribir texto en un archivo nuevo o sobrescribir si existe
        File.WriteAllText(nombreArchivo, texto);

        // Leer todo el texto del archivo
        string contenido = File.ReadAllText(nombreArchivo);

        /////////////////////////////////
        // SALIDA DE DATOS EN PANTALLA //
        /////////////////////////////////

        // Mostrar en consola los números en la lista de flotantes
        Console.WriteLine("Números en la lista de flotantes:");
        foreach (float numero in listaDeFlotantes)
        {
            Console.WriteLine(numero);
        }

        // Mostrar números mayores que cinco
        Console.WriteLine("\nNúmeros mayores que cinco:");
        foreach (float numero in mayoresQueCinco)
        {
            Console.WriteLine(numero);
        }

        // Mostrar precios en el diccionario
        Console.WriteLine("\nPrecios en el diccionario:");
        foreach (KeyValuePair<string, float> par in diccionarioDePrecios)
        {
            Console.WriteLine($"Producto: {par.Key}, Precio: {par.Value}");
        }

        // Mostrar el contenido leído del archivo de texto
        Console.WriteLine("\nContenido del archivo de texto:");
        Console.WriteLine(contenido);
    }
}

File: cursos/pria/src/python/7_bibliotecas.py
"""
Este script demuestra el uso de varias librerías estándar de Python para realizar operaciones matemáticas,
generar números aleatorios, trabajar con colecciones, manipular archivos y mostrar información en pantalla.
"""

# Importar módulos necesarios
import math      # Para funciones matemáticas avanzadas
import random    # Para generar números aleatorios
import os        # Para operaciones del sistema, como verificar si un archivo existe
import statistics  # Para cálculos estadísticos
import itertools   # Para combinaciones y permutaciones
from collections import Counter  # Para contar elementos

#########################################
# EJEMPLOS DE USAR LIBRERÍAS INTERNAS   #
#########################################

# MATEMÁTICAS (USANDO EL MÓDULO math)

# Variables de punto flotante
x = 5.0
y = 2.0

# Constantes matemáticas
pi = math.pi       # Constante π
e = math.e         # Constante e (Número de Euler)

print(f"Constante pi: {pi}")
print(f"Constante e: {e}")

# Operaciones matemáticas
potencia = math.pow(x, y)     # x elevado a la potencia y
exponencial = math.exp(x)     # e elevado a x
logaritmo = math.log(x)       # Logaritmo natural de x

print(f"{x} elevado a {y} es {potencia}")
print(f"e elevado a {x} es {exponencial}")
print(f"Logaritmo natural de {x} es {logaritmo}")

# Valor absoluto
valor_absoluto = abs(-x)
print(f"Valor absoluto de {-x} es {valor_absoluto}")

# Redondeo y funciones de piso y techo
redondeado = round(x)
piso = math.floor(x)
techo = math.ceil(x)

print(f"{x} redondeado es {redondeado}")
print(f"El piso de {x} es {piso}")
print(f"El techo de {x} es {techo}")

# Funciones trigonométricas (ángulos en radianes)
seno = math.sin(x)
coseno = math.cos(x)
tangente = math.tan(x)
arco_tangente = math.atan(x)
arco_tangente2 = math.atan2(y, x)

print(f"Seno de {x} es {seno}")
print(f"Coseno de {x} es {coseno}")
print(f"Tangente de {x} es {tangente}")
print(f"Arcotangente de {x} es {arco_tangente}")
print(f"Arcotangente2 de ({y}, {x}) es {arco_tangente2}")

# Máximo y mínimo
maximo = max(x, y)
minimo = min(x, y)

print(f"El máximo entre {x} y {y} es {maximo}")
print(f"El mínimo entre {x} y {y} es {minimo}")

# ALEATORIEDAD (USANDO EL MÓDULO random)

# Número flotante aleatorio entre 0 y 1
numero_aleatorio = random.random()
print(f"\nNúmero aleatorio entre 0 y 1 → {numero_aleatorio}")

# Número entero aleatorio entre dos valores, inclusivo
min_val = -3
max_val = 10
entero_aleatorio = random.randint(min_val, max_val)
print(f"Número entero aleatorio entre {min_val} y {max_val} → {entero_aleatorio}")

# Número flotante aleatorio entre dos valores
numero_aleatorio_flotante = random.uniform(min_val, max_val)
print(f"Número flotante aleatorio entre {min_val} y {max_val} → {numero_aleatorio_flotante}")

# COLECCIONES

# Lista de números de punto flotante
lista_de_flotantes = [1.5, 2.3, 3.7, 4.2, 5.9]

# Añadir elementos a la lista
lista_de_flotantes.append(6.1)
lista_de_flotantes.extend([7.4, 8.6, 9.0, 10.3])

print("\nLista de flotantes:")
print(lista_de_flotantes)

# Diccionario con claves de tipo string y valores de tipo float
diccionario_de_precios = {
    "Manzana": 0.75,
    "Pan": 1.50,
    "Leche": 2.99
}

# Añadir elementos al diccionario
diccionario_de_precios["Huevos"] = 1.99
diccionario_de_precios.update({"Queso": 3.49})

print("\nDiccionario de precios:")
for producto, precio in diccionario_de_precios.items():
    print(f"{producto} → {precio} €\t", end='')

# FILTRADO Y ORDENACIÓN (USANDO COMPRENSIONES DE LISTAS Y sorted)

# Filtrar elementos: obtener todos los números mayores que 5
mayores_que_cinco = [numero for numero in lista_de_flotantes if numero > 5.0]
print("\n\nNúmeros mayores que cinco:")
print(mayores_que_cinco)

# ESCRIBIR, LEER Y MOSTRAR ARCHIVOS DE TEXTO

nombre_archivo = 'texto.txt'
texto = "¡Hola, mundo!\n¡Esta es otra línea de texto!"

# Escribir texto en un archivo nuevo o sobrescribir si existe
with open(nombre_archivo, 'w', encoding='utf-8') as archivo:
    archivo.write(texto)

# Verificar si el archivo existe antes de leer
if os.path.exists(nombre_archivo):
    # Leer todo el texto del archivo
    with open(nombre_archivo, 'r', encoding='utf-8') as archivo:
        contenido = archivo.read()
else:
    contenido = ""
    print(f"El archivo {nombre_archivo} no existe.")

# Mostrar el contenido leído del archivo de texto
print("\nContenido del archivo de texto:")
print(contenido)


# DEMOSTRACIÓN DE OTRAS FUNCIONALIDADES

# Uso del módulo statistics para calcular estadísticas básicas
media = statistics.mean(lista_de_flotantes)
mediana = statistics.median(lista_de_flotantes)
desviacion = statistics.stdev(lista_de_flotantes)

print("\nEstadísticas de la lista de flotantes:")
print(f"Media: {media}\t", end='')
print(f"Mediana: {mediana}\t", end='')
print(f"Desviación estándar: {desviacion}")

# Uso del módulo collections para contar elementos
# Contar frecuencias de números redondeados
conteo_numeros = Counter(round(num) for num in lista_de_flotantes)
print("\nFrecuencia de números redondeados:")
for numero, frecuencia in conteo_numeros.items():
    print(f"Número: {numero}, Frecuencia: {frecuencia}; ", end='')

# Uso de itertools para combinaciones y permutaciones
# Obtener todas las combinaciones de 2 elementos de la lista de flotantes
combinaciones = list(itertools.combinations(lista_de_flotantes, 2))
print("\n\nCombinaciones de 2 elementos de la lista de flotantes:")
print(combinaciones)

# Obtener todas las permutaciones de 2 elementos de la lista de flotantes
permutaciones = list(itertools.permutations(lista_de_flotantes, 2))
print("\nPermutaciones de 2 elementos de la lista de flotantes:")
print(permutaciones)
